File: Python Submissions/Longest-Palindrome.py
# 5. Longest Palindromic Substring
# Given a string s, return the longest palindromic substring in s.

class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        T = "^#" + ("#".join(list(s))) + "#$"
        n = len(T)
        P = [0] * n
        C = 0
        R = 0

        for i in range(1, n-1):
            P[i] = min(R - i, P[2*C - i]) if (R > i) else 0
            while T[i + 1 + P[i]] == T[i - 1 - P[i]]:
                P[i] += 1

                if i + P[i] > R:
                    C = i
                    R = i + P[i]      

        max_len = max(P)
        center_index = P.index(max_len)
        return s[(center_index - max_len) // 2 : (center_index + max_len) // 2]
    
solution = Solution()
print(solution.longestPalindrome("babad")) # "aba" or "bab"

File: Java Submissions/Longest-Palindrome.java
// 5. Longest Palindromic Substring
// Given a string s, return the longest palindromic substring in s.

class Solution {
    public String longestPalindrome(String s) {
        String[] T = ("^#" + String.join("#", s.split("")) + "#$").split("");
        int n = T.length;
        int[] P = new int[n];
        int C = 0; 
        int R = 0;
        int max_len = 0;
        int center_index = 0;

        for (int i = 1; i < n-1; i++) {
            P[i] = (R > i) ? Math.min(R - i, P[2*C - i]) : 0;
            while (T[i + 1 + P[i]].equals(T[i - 1 - P[i]])) {
                P[i]++;
                if (i + P[i] > R) {
                    C = i;
                    R = i + P[i];
                }                
            }
            max_len = Math.max(max_len, P[i]);
            if (max_len == P[i]) {
                center_index = i;
            }
        }
        return s.substring((center_index - max_len) / 2, (center_index + max_len) / 2);
    }
}

public static void main(String[] args) {
    Solution solution = new Solution();
    System.out.println(solution.longestPalindrome("babad")); // "aba" or "bab"
}